File: apps/cargo-tracker/test/cargotracker/shipper/domain/model/aggregates/ShipperSpec.scala
package cargotracker.shipper.domain.model.aggregates

import cargotracker.shared.domain.{ShipperId, ShipperType}
import cargotracker.shipper.domain.model.valueobjects.DiscountRate
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
class ShipperSpec extends AnyFunSuite with Matchers:

  private val validId = ShipperId("SH-000001").toOption.get

  test("個人荷主を生成できる"):
    val res = Shipper.individual(
      shipperId = validId,
      name = "山田太郎",
      email = "yamada@example.com",
      phone = "03-1234-5678",
      address = "東京都千代田区..."
    )
    res.isRight shouldBe true
    val shipper = res.toOption.get
    shipper.shipperType shouldBe ShipperType.Individual
    shipper.discountRate shouldBe DiscountRate.zero

  test("法人荷主は契約番号と割引率を持つ"):
    val res = Shipper.corporate(
      shipperId = validId,
      name = "株式会社 ABC",
      email = "sales@abc.co.jp",
      phone = "03-9999-9999",
      address = "東京都港区...",
      contractNumber = "CT-2026-0001",
      discountRate = DiscountRate(0.15).toOption.get
    )
    res.isRight shouldBe true
    val shipper = res.toOption.get
    shipper.shipperType shouldBe ShipperType.Corporate
    shipper.contractNumber shouldBe Some("CT-2026-0001")
    shipper.discountRate.value shouldBe 0.15

  test("名前が空白だけの荷主は拒否される"):
    Shipper.individual(validId, "   ", "u@x.com", "0", "addr") shouldBe Left(
      Shipper.InvalidName
    )

  test("メール形式が不正な荷主は拒否される"):
    Shipper.individual(
      validId,
      "山田",
      "invalid",
      "0",
      "addr"
    ) shouldBe Left(Shipper.InvalidEmail)

  test("DiscountRate は 0〜0.30 の範囲のみ受け付ける"):
    DiscountRate(0.0).map(_.value) shouldBe Right(0.0)
    DiscountRate(0.30).map(_.value) shouldBe Right(0.30)
    DiscountRate(0.31) shouldBe Left(DiscountRate.OutOfRange)
    DiscountRate(-0.01) shouldBe Left(DiscountRate.OutOfRange)

  test("ShipperType の表記が荷主インスタンスから取り出せる"):
    val individual = Shipper
      .individual(validId, "Y", "y@x.com", "0", "addr")
      .toOption
      .get
    individual.shipperType shouldBe ShipperType.Individual

File: docs/article/source/haskell/apps/cargo-tracker/test/unit/Shipper/Domain/Model/ShipperSpec.hs
{- | Shipper 集約のテスト (IT1 US02/US03 2.1)

US02 個人荷主 / US03 法人荷主は同一集約ルート Shipper の sum type で表現する。
法人は CorporateNumber (13 桁) と ContractRank (Bronze/Silver/Gold) を持つ。
-}
module Shipper.Domain.Model.ShipperSpec (spec) where

import qualified Data.Text as T
import Test.Hspec

import Cargotracker.Shared.Domain.DomainError (DomainError (..))
import Cargotracker.Shipper.Domain.Model.Shipper
  ( ContractRank (..),
    CorporateNumber (..),
    Shipper (..),
    ShipperKind (..),
    discountPercentage,
    mkCorporateNumber,
    mkCorporateShipper,
    mkIndividualShipper,
  )
import Cargotracker.Shipper.Domain.Model.Value.Address (Address (..), mkAddress)
import Cargotracker.Shipper.Domain.Model.Value.ContactEmail (ContactEmail (..), mkContactEmail)
import Cargotracker.Shipper.Domain.Model.Value.ShipperId (ShipperId (..), mkShipperId)
import Cargotracker.Shipper.Domain.Model.Value.ShipperName (ShipperName (..), mkShipperName)

validId :: ShipperId
validId =
  case mkShipperId "SHP-A1B2C3" of
    Right s -> s
    Left _ -> error "test setup: invalid id"

validEmail :: ContactEmail
validEmail =
  case mkContactEmail "shipper@example.com" of
    Right e -> e
    Left _ -> error "test setup: invalid email"

validAddress :: Address
validAddress =
  case mkAddress "東京都港区芝公園 4-2-8" of
    Right a -> a
    Left _ -> error "test setup: invalid address"

validName :: ShipperName
validName =
  case mkShipperName "山田 太郎" of
    Right n -> n
    Left _ -> error "test setup: invalid name"

spec :: Spec
spec = do
  describe "ShipperId" $ do
    it "SHP- プレフィックスがないと不正" $
      mkShipperId "A1B2C3" `shouldBe` Left (InvalidShipperId "expected SHP-XXXXXX")
    it "プレフィックス後が 6 文字でないと不正" $
      mkShipperId "SHP-ABC" `shouldBe` Left (InvalidShipperId "expected SHP-XXXXXX")
    it "正しいフォーマットは構築できる" $
      mkShipperId "SHP-A1B2C3" `shouldBe` Right (ShipperId "SHP-A1B2C3")

  describe "Address" $ do
    it "空文字は不正" $
      mkAddress "" `shouldBe` Left (InvalidShipperId "empty address")
    it "500 文字超は不正" $
      let long = T.replicate 501 "あ"
       in mkAddress long `shouldBe` Left (InvalidShipperId "address too long (max 500)")
    it "500 文字は構築できる" $
      let just500 = T.replicate 500 "あ"
       in mkAddress just500 `shouldBe` Right (Address just500)

  describe "ContactEmail" $ do
    it "@ なしは不正" $
      mkContactEmail "noatsign" `shouldBe` Left (InvalidShipperId "invalid email")
    it "有効なメールは構築できる" $
      mkContactEmail "x@y.z" `shouldBe` Right (ContactEmail "x@y.z")

  describe "CorporateNumber" $ do
    it "13 桁ではないと不正" $
      mkCorporateNumber "12345" `shouldBe` Left (InvalidShipperId "expected 13 digits")
    it "数字以外を含むと不正" $
      mkCorporateNumber "12345678901AB" `shouldBe` Left (InvalidShipperId "digits only")
    it "13 桁の数字は構築できる" $
      mkCorporateNumber "1234567890123" `shouldBe` Right (CorporateNumber "1234567890123")

  describe "ContractRank" $ do
    it "Bronze/Silver/Gold の 3 段階" $
      [Bronze, Silver, Gold] `shouldBe` [minBound .. maxBound]

  describe "mkIndividualShipper (US02)" $ do
    it "ID/氏名/メール/住所で構築でき kind は Individual" $ do
      let s = mkIndividualShipper validId validName validEmail validAddress
      shipperId s `shouldBe` validId
      shipperName s `shouldBe` validName
      shipperKind s `shouldBe` Individual

  describe "mkCorporateShipper (US03)" $ do
    it "ID/社名/メール/住所/法人番号/契約ランクで構築でき kind は Corporate" $ do
      Right cn <- pure (mkCorporateNumber "1234567890123")
      let s =
            mkCorporateShipper
              validId
              validName
              validEmail
              validAddress
              cn
              Gold
      shipperKind s `shouldBe` Corporate cn Gold

  describe "mkShipperName (T-09)" $ do
    it "空文字は Left" $
      mkShipperName "" `shouldBe` Left (InvalidShipperId "shipper name must not be empty")
    it "256 文字超は Left" $
      mkShipperName (T.replicate 256 "a")
        `shouldBe` Left (InvalidShipperId "shipper name too long (max 255)")
    it "前後空白は trim される" $
      mkShipperName "  山田 太郎  " `shouldBe` Right (ShipperName "山田 太郎")

  describe "discountPercentage (US22, IT7)" $ do
    it "Individual は 0%" $ do
      let s = mkIndividualShipper validId validName validEmail validAddress
      discountPercentage s `shouldBe` 0
    it "Corporate Bronze は 5%" $ do
      Right cn <- pure (mkCorporateNumber "1234567890123")
      let s = mkCorporateShipper validId validName validEmail validAddress cn Bronze
      discountPercentage s `shouldBe` 5
    it "Corporate Silver は 10%" $ do
      Right cn <- pure (mkCorporateNumber "1234567890123")
      let s = mkCorporateShipper validId validName validEmail validAddress cn Silver
      discountPercentage s `shouldBe` 10
    it "Corporate Gold は 15%" $ do
      Right cn <- pure (mkCorporateNumber "1234567890123")
      let s = mkCorporateShipper validId validName validEmail validAddress cn Gold
      discountPercentage s `shouldBe` 15
